lib/ansible/modules/windows/win_http_proxy.ps1

```
@@ -0,0 +1,264 @@
+                    if (!String.IsNullOrEmpty(ProxyBypass))
+                        proxyInfo.lpszProxyBypass = Marshal.StringToHGlobalUni(ProxyBypass);
+                }
+
+                if (!NativeMethods.WinHttpSetDefaultProxyConfiguration(proxyInfo))
+                    throw new Win32Exception("WinHttpSetDefaultProxyConfiguration() failed");
+            }
+        }
+
+        public void Refresh()
+        {
+            using (NativeHelpers.WINHTTP_PROXY_INFO proxyInfo = new NativeHelpers.WINHTTP_PROXY_INFO())
+            {
+                if (!NativeMethods.WinHttpGetDefaultProxyConfiguration(proxyInfo))
+                    throw new Win32Exception("WinHttpGetDefaultProxyConfiguration() failed");
+
+                Proxy = Marshal.PtrToStringUni(proxyInfo.lpszProxy);
+                ProxyBypass = Marshal.PtrToStringUni(proxyInfo.lpszProxyBypass);
+            }
+        }
+
+        public static WinINetProxy GetIEProxyConfig()
+        {
+            using (NativeHelpers.WINHTTP_CURRENT_USER_IE_PROXY_CONFIG ieProxy = new NativeHelpers.WINHTTP_CURRENT_USER_IE_PROXY_CONFIG())
+            {
+                if (!NativeMethods.WinHttpGetIEProxyConfigForCurrentUser(ieProxy))
+                    throw new Win32Exception("WinHttpGetIEProxyConfigForCurrentUser() failed");
+
+                return new WinINetProxy
+                {
+                    AutoDetect = ieProxy.fAutoDetect,
+                    AutoConfigUrl = Marshal.PtrToStringUni(ieProxy.lpszAutoConfigUrl),
+                    Proxy = Marshal.PtrToStringUni(ieProxy.lpszProxy),
+                    ProxyBypass = Marshal.PtrToStringUni(ieProxy.lpszProxyBypass),
+                };
+            }
+        }
+    }
+}
+'@
+Add-CSharpType -References $win_http_invoke -AnsibleModule $module
+
+$actual_proxy = New-Object -TypeName Ansible.WinHttpProxy.WinHttpProxy
+
+$module.Diff.before = @{
+    proxy = $actual_proxy.Proxy
+    bypass = $actual_proxy.ProxyBypass
+}
+
+if ($source -eq "ie") {
```

IIRC, the underlying "import from IE" logic provided by Microsoft silently ignores the autoconfig URL -- which seems to be a mistake to me. Do we want to propagate that behavior? I suspect a warning would be better.

---

(Admittedly, I haven't actually looked at that in a very long time, so I could be mis-remembering.)

---

I couldn't find any way to actually set an auto config url in netsh/winhttp so I think it is a limitation of the latter? Could be wrong but the [WINHTTP_PROXY_INFO](https://docs.microsoft.com/en-us/windows/desktop/api/Winhttp/ns-winhttp-__unnamed_struct_3) structure does not have a field for this.

This sounds like more of a documentation thing with the potential of adding this in as a separate param if it is possible.

---

Sorry, I guess I wasn't clear. The "mistake" I was referring to was the fact that netsh doing this type of import **_silently_** ignores the auto-config URL (IIRC).

I'm not surprised that there's no winhttp option for this, since that would imply winhttp.dll having a JS interpreter (which doesn't match well with its use-case). From that perspective, _ignoring_ the auto-config URL when doing an import is entirely reasonable on MS' part -- **_except_** if it is done silently. IMHO that should really be a warning.

---

Hmm I might be missing something, it's early morning here. Just to double check you are saying that the netsh import doesn't use the auto config value at all? That's the assumption that I was under. The trouble with adding warnings to modules is that it will automatically display when being used. This case is more suited towards documentation where people reading the options should be aware that `source: ie` will just import the proxy URL and bypass list.

---

RE not used at all -- correct. (I'm still going from memory on that one, but that is how I remember it.) More specifically, if you import from IE, which _only_ is set to use a [valid] auto-config script, you'll get nothing whatsoever in the winhttp settings.

As far warning or doc, my _preference_ would be to do both, but I won't get up on a high horse about it either way. :-)

---

(Should probably actually double-check my recollection on this first though. We'd all feel really silly if I'm remembering wrong.)

---

Cool so yes the import is only for explicit proxy URLs and bypass list. I'll update the docs to further clarify this but I don't think a warning is necessary. This is up to the user making sure they understand the behaviour and read the docs.

---

Good enough for me.

---

I also found [WinHttpGetProxyForUrl](https://docs.microsoft.com/en-us/windows/desktop/api/winhttp/nf-winhttp-winhttpgetproxyforurl) which seems to be the mechanism for WinHTTP applications to support auto config urls. This indicates it's up to the application to look this up and set on their requests and not a system wide setting.

---

Yeah, your second link is pretty succinct on the matter:
> AutoProxy support is not fully integrated into the HTTP stack in WinHTTP. Before sending a request, the application must call ...

In that respect, your doc content changes look good.